Week 4 Day 5 update: web components (#74)
microsoft/win-student-devs: docs/30DaysOfPWA/platforms-practices/05.md

# 4.5: PWA Maintainability: Using Web Components

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

Welcome to Week 4, Day 5 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.

![Placeholder Banner Only. Replace when final assets ready.](_media/day5.png)

## Where Do We Start With Maintainability?
Keeping our PWA maintainable, or any application for that matter, can be a serious undertaking with tons of unique challenges. The web capabilities that make PWAs possible allows for more capable and complicated web experiences, but it comes with a price. How do we keep our increasingly complex PWA projects clean and extensible?

> There's a ton of options, but for the sake of simplicity, we'll just be looking at one technology we can leverage in maintainable projects: **web components!**

Reviewer: If any additional content is desired for this section on "why web components", the FAST site has some points you can feel free to borrow: https://www.fast.design/docs/resources/why-web-components

Let's take a look at what makes web components so awesome.

## Encapsulation with Shadow DOM

When we create a web component, all of our behavior, html, and styling is contained to that component. We don't have to worry about it interfering with other aspects of our project when we include our component in our application. This a huge bonus for maintainability: it allows us to segment our code and avoid overly-verbose spaghetti code in our project files. Additionally, for larger PWAs that may have multiple developers working at once, web commponents allow for concurrent contribution by keeping unrelated components separated from each other.

Web components make this possible using [**Shadow DOM**](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM#high-level_view), which allows our component to live in a "hidden" portion of our larger DOM tree. We can prevent clashing with the rest of our DOM by isolating the portion related to the web component. 

## Web Standard

> Web Components are a **Web Standard**, which means they are considered a fundamental model for component behavior on the Web. This means we can rely on the web components we use and build to work just about everywhere:

Reviewer 2: A bit confused by this sentence: "This means we can rely on the web components we use and build to work just about everywhere". Is there a verb missing somewhere?

![Compatability list for web components](_media/component-web-standards.png)

As far as maintainability goes for our PWAs, this is great. We know our components will work in different browsers, and we don't have to write extra code to keep our project functional. When we go to update or extend one of our components, we can work freely without worrying about compatability issues.

## Building Web Components with FAST

> There are lots of framework options for building web components, but today we'll be using [the FAST framework](https://www.fast.design/docs/introduction) to see a beginner example.

Reviewer 2: Is it worth mentioning here that you can also use "vanilla" web components too? Like, you don't need a framework to get started. And depending on the complexity of your organization and project, it might not make sense to use one at all.

Reviewer: If we do add that, I think a good idea would be to point out that FAST and Lit aren't like typical JavaScript frameworks in that they are very small and fully interoperate through standards. So, the concerns of "choosing a framework" or event "vanilla js" are not quite the same as people may be used to.

Let's say we are building a naive `<movie-review>` element that will live in an `MovieReview.ts` file.

First, we can start with an element with a few attributes:

```typescript
import { FASTElement, customElement, attr } from '@microsoft/fast-element';

@customElement('movie-review')
export class MovieReview extends FASTElement {

>     @attr name: string = "No Movie Listed";

Reviewer: Might want to avoid using `name` since that has other uses in HTML.

    @attr rating: string = "0";
}
```

Next, we can provide an HTML template so we can display our attributes:

```typescript
import { FASTElement, customElement, attr, html } from '@microsoft/fast-element';

// our html template for our component
const template = html<MovieReview>`
    <div class="reviewDiv">
        <div class="nameDiv">${x => x.name}</div>
        <div class="ratingDiv">${x=> x.rating} stars!<div>
    </div>
`

// that we can pass into our customElement
@customElement({
    name: 'movie-review',
    template
})
export class MovieReview extends FASTElement {
    @attr name: string = "No Movie Listed";
    @attr rating: string = "0";
}
```

We could then use our element like so:
```html

> <movie-review movieName="Fast and Furious: Tokyo Drift" rating="5">

Reviewer: There's a bug in the code sample here. `movieName` isn't an attribute on the custom element as defined. Also, HTML attributes are case normalized by the browser so you'd want this to be `moviename` to avoid confusion.

```

This is just going to give us some boring black text. Let's see how we can add styling:

```typescript
import { FASTElement, customElement, attr, html, css } from '@microsoft/fast-element';

const styles = css`
    .reviewDiv {
        border: 1px solid black;
        display: inline-block;
        padding: 10px;
    }

    .nameDiv { 
        font-size: 24; 
        font-weight: bold;
        font-style: italic;
    }

    .ratingDiv { 
        color: blue; 
    }
`
```

And we can pass those styles in just like our template:

```typescript
@customElement({
    name: 'movie-review',
    template,
    styles
})
export class MovieReview extends FASTElement {
    @attr name: string = "No Movie Listed";
    @attr rating: string = "0";
}
```

Our _beautiful_ component would then look something like this:

> ![](_media/movie-review.png)

Reviewer 2: Missing alt text here.

Author: ```suggestion
![Image of our movie review web component](_media/movie-review.png)
```

We now have a reuseable `<movie-review>` component. As you can see, this example is fairly naive, but you can start to imagine the possibilites. Let's say we had a more in-depth `<movie-review>` component that we could use inside a related `<movie-review-list>` component. We could then change the appearance and functionality of our movie reviews separate from the list that displays them, allowing us extend and update our `<movie-review>` component regarldess of the environment it lives in. Encapsulated and maintainable!

> If you want more depth on building quality components and adding functionality, check out the [FAST documentation](https://www.fast.design/docs/introduction).

Reviewer: ```suggestion
If you want more depth on building quality components and adding functionality, or if you want to leverage the dozens of pre-built components FAST offers, check out the [FAST documentation](https://www.fast.design/docs/introduction).
```

I thought it might be nice to squeeze a bit in there about the other things FAST offers besides just the library for building components.

> FAST isn't the only tool for building web components. For example, the PWA starter uses [lit](https://lit.dev/docs/), which is another web components framework.

Reviewer 3: You could link here to all the ways you can create a web component - developers can find the framework that is right for them https://webcomponents.dev/blog/all-the-ways-to-make-a-web-component/

# 4.5: PWA Maintainability: Using Web Components

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

> Welcome to Week 4, Day 5 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.

Reviewer 4: ```suggestion
Welcome to Week 4, Day 5 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilities they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.
```

> ![Placeholder Banner Only. Replace when final assets ready.](_media/day5.png)

Reviewer 4: ```suggestion
![image of title and author.](_media/day5.png)
```

## Where Do We Start With Maintainability?
Keeping our PWA maintainable, or any application for that matter, can be a serious undertaking with tons of unique challenges. The web capabilities that make PWAs possible allow for more capable and complicated web experiences, but it comes with a price. How do we keep our increasingly complex PWA projects clean and extensible?

Out of the many options available to us, today we'll be looking at one technology we can leverage in our projects: **web components!**

## Web Components?
Web components are set of web technologies that allow you to create custom standalone components (like your very own `<custom-button>`, for example) that work anywhere on the web. They come with a whole suite of development advantages, including project maintainability. Check out this ["Why Web Components?"](https://aka.ms/learn-pwa/30days-4.5/fast.design/docs/resources/why-web-components) article for a quick primer on what Web Components bring to the table.

Now, let's take a look at how web components can improve our PWAs and make our projects more maintainable.

## Encapsulation with Shadow DOM

> When we create a web component, all of our behavior, html, and styling is contained to that component. We don't have to worry about it interfering with other aspects of our project when we include our component in our application. This a huge bonus for maintainability: it allows us to segment our code and avoid overly-verbose spaghetti code in our project files. Additionally, for larger PWAs that may have multiple developers working at once, web commponents allow for concurrent contribution by keeping unrelated components separated from each other.

Reviewer 4: ```suggestion
When we create a web component, all of our behavior, html, and styling is contained to that component. We don't have to worry about it interfering with other aspects of our project when we include our component in our application. This a huge bonus for maintainability: it allows us to segment our code and avoid overly-verbose spaghetti code in our project files. Additionally, for larger PWAs that may have multiple developers working at once, web components allow for concurrent contribution by keeping unrelated components separated from each other.
```

Web components make this possible using [**Shadow DOM**](https://aka.ms/learn-pwa/30days-4.5/developer.mozilla.org/docs/Web/Web_Components/Using_shadow_DOM), which allows our component to live in a "hidden" portion of our larger DOM tree. We can prevent clashing with the rest of our DOM by isolating the portion related to the web component. 

## Web Standard
Web Components are a **Web Standard**, which means they are considered a fundamental model for component behavior on the Web. This means we can rely on the web components we build to work just about everywhere:

> ![Compatability list for web components](_media/component-web-standards.png)

Reviewer 4: ```suggestion
![Compatibility list for web components](_media/component-web-standards.png)
```

> Another benefit to this is that web component frameworks are generally small and fuly interoperable. This means that you can use web components within larger modern frameworks (like React or Angular) and they will still work as expected. You can even ditch the frameworks all together and build web components in vanilla javascript.

Reviewer 4: ```suggestion
Another benefit to this is that web component frameworks are generally small and fully interoperable. This means that you can use web components within larger modern frameworks (like React or Angular) and they will still work as expected. You can even ditch the frameworks all together and build web components in vanilla javascript.
```

> As far as maintainability goes for our PWAs, this is great. We know our components will work in different browsers and frameworks, and we don't have to write extra code to keep our project functional. When we go to update or extend one of our components, we can work freely without worrying about compatability issues. Web Standards also generally guarantee forward-compability: our components will still work as the open web grows and changes.

Reviewer 4: ```suggestion
As far as maintainability goes for our PWAs, this is great. We know our components will work in different browsers and frameworks, and we don't have to write extra code to keep our project functional. When we go to update or extend one of our components, we can work freely without worrying about compatibility issues. Web Standards also generally guarantee forward-compatibility: our components will still work as the open web grows and changes.
```

> Web Components are a **Web Standard**, which means they are considered a fundamental model for component behavior on the Web. This means we can rely on the web components we build to work just about everywhere:
> 
> ![Compatability list for web components](_media/component-web-standards.png)
> 
> Another benefit to this is that web component frameworks are generally small and fuly interoperable. This means that you can use web components within larger modern frameworks (like React or Angular) and they will still work as expected. You can even ditch the frameworks all together and build web components in vanilla javascript.
> 
> As far as maintainability goes for our PWAs, this is great. We know our components will work in different browsers and frameworks, and we don't have to write extra code to keep our project functional. When we go to update or extend one of our components, we can work freely without worrying about compatability issues. Web Standards also generally guarantee forward-compability: our components will still work as the open web grows and changes.
> 

Reviewer 4: ```suggestion
Web Components are a **Web Standard**, which means they are considered a fundamental model for component behavior on the Web. This means we can rely on the web components we build to work just about everywhere:

![Compatibility list for web components](_media/component-web-standards.png)

Another benefit to this is that web component frameworks are generally small and fully interoperable. This means that you can use web components within larger modern frameworks (like React or Angular) and they will still work as expected. You can even ditch the frameworks all together and build web components in vanilla javascript.

As far as maintainability goes for our PWAs, this is great. We know our components will work in different browsers and frameworks, and we don't have to write extra code to keep our project functional. When we go to update or extend one of our components, we can work freely without worrying about compatibility issues. Web Standards also generally guarantee forward-compatibility: our components will still work as the open web grows and changes.

```

    name: 'movie-review',
    template
})
export class MovieReview extends FASTElement {
    @attr movieName: string = "No Movie Listed";
    @attr rating: string = "0";
}
```

We could then use our element like so:
```html
<movie-review moviename="Fast and Furious: Tokyo Drift" rating="5">
```

This is just going to give us some boring black text. Let's see how we can add styling:

```typescript
import { FASTElement, customElement, attr, html, css } from '@microsoft/fast-element';

const styles = css`
    .reviewDiv {
        border: 1px solid black;
        display: inline-block;
        padding: 10px;
    }

    .nameDiv { 
        font-size: 24; 
        font-weight: bold;
        font-style: italic;
    }

    .ratingDiv { 
        color: blue; 
    }
`
```

And we can pass those styles in just like our template:

```typescript
@customElement({
    name: 'movie-review',
    template,
    styles
})
export class MovieReview extends FASTElement {
    @attr movieName: string = "No Movie Listed";
    @attr rating: string = "0";
}
```

Our _beautiful_ component would then look something like this:

![Image of our movie review web component](_media/movie-review.png)

> We now have a reuseable `<movie-review>` component. As you can see, this example is fairly naive, but you can start to imagine the possibilites. Let's say we had a more in-depth `<movie-review>` component that we could use inside a related `<movie-review-list>` component. We could then change the appearance and functionality of our movie reviews separate from the list that displays them, allowing us extend and update our `<movie-review>` component regarldess of the environment it lives in. Encapsulated and maintainable!

Reviewer 4: ```suggestion
We now have a reusable `<movie-review>` component. As you can see, this example is fairly naive, but you can start to imagine the possibilities. Let's say we had a more in-depth `<movie-review>` component that we could use inside a related `<movie-review-list>` component. We could then change the appearance and functionality of our movie reviews separate from the list that displays them, allowing us extend and update our `<movie-review>` component regardless of the environment it lives in. Encapsulated and maintainable!
```
